compiler/src/iree/compiler/Dialect/Util/Transforms/VerifyInitializationOrder.cpp

```diff
@@ -0,0 +1,407 @@
+// Copyright 2025 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Dialect/Util/Analysis/GlobalTable.h"
+#include "iree/compiler/Dialect/Util/IR/UtilOps.h"
+#include "iree/compiler/Dialect/Util/Transforms/Passes.h"
+#include "mlir/Analysis/CallGraph.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/Interfaces/ControlFlowInterfaces.h"
+#include "mlir/Pass/Pass.h"
+
+namespace mlir::iree_compiler::IREE::Util {
+
+#define GEN_PASS_DEF_VERIFYINITIALIZATIONORDERPASS
+#include "iree/compiler/Dialect/Util/Transforms/Passes.h.inc"
+
+namespace {
+
+// Tracks which functions are only reachable from initializers.
+struct InitializerReachability {
+  // Functions that are only callable from initializers (not from external).
+  DenseSet<Operation *> initializerOnlyFunctions;
+  // Functions reachable from external entry points.
+  DenseSet<Operation *> externallyReachableFunctions;
+};
+
+class VerifyInitializationOrderPass
+    : public impl::VerifyInitializationOrderPassBase<
+          VerifyInitializationOrderPass> {
+public:
+  void runOnOperation() override {
+    auto moduleOp = getOperation();
+    GlobalTable globalTable(moduleOp);
+    globalTable.rebuild();
+
+    // Collect initialization points in module order.
+    SmallVector<IREE::Util::InitializerOpInterface> initializerOps;
+    DenseMap<Operation *, size_t> opOrdinalMap;
```

Comment:
I feel that either they should belong to class members, so you don't pass them to member functions. Or the member functions can be static functions. I'm +1 on the former one.